ark/utils/metacluster_remap_gui/metaclustergui.py

```diff
@@ -0,0 +1,463 @@
+import warnings
+
+import ipywidgets as widgets
+import matplotlib.patheffects as path_effects
+import matplotlib.pyplot as plt
+import numpy as np
+import seaborn as sns
+from IPython.display import display
+from scipy.cluster.hierarchy import dendrogram
+from scipy.stats import zscore
+from scipy.stats.stats import F_onewayBadInputSizesWarning
+
+from .colormap_helper import distinct_cmap
+from .metaclusterdata import MetaClusterData
+from .throttle import throttle
+from .zscore_norm import ZScoreNormalize
+
+# Third party ipympl causing this in it's backend_agg startup code
+warnings.filterwarnings("ignore", message="nbagg.transparent is deprecated")
+
+DEBUG_VIEW = widgets.Output(layout={'border': '1px solid black'})
+
+
+class MetaClusterGui():
+    def __init__(self, metaclusterdata, heatmapcolors=sns.diverging_palette(240, 10, n=3, as_cmap=True), width=17.0, debug=False, enable_throttle=True):  # noqa
+        self.width: float = width
+        self.heatmapcolors: str = heatmapcolors
+        self.mcd: MetaClusterData = metaclusterdata
+        self.selected_clusters = set()
+
+        self.make_gui()
+
+        if debug:
+            self.enable_debug_mode()
+
+        if enable_throttle:
+            throttler = throttle(.3)
+            self.update_gui = throttler(self.update_gui)
+
+    @property
+    def max_zscore(self):
+        return self.zscore_clamp_slider.value
+
+    @property
+    def cmap(self):
+        # will never have more metaclusters than clusters
+        return distinct_cmap(self.mcd.cluster_count)
+
+    def preplot(self, df):
+        return df.apply(zscore).clip(upper=self.max_zscore).T
+
+    def make_gui(self):
```

Comment:
This function is 126 lines: is there a reason we can't break this up into helper functions?

Reply:
No i'll break it up. It could also be better factored in terms of "single level of abstraction' principal.

Right now it is a mixture of high and low level. (make_widget, update_gui, vs everything else)
![image](https://user-images.githubusercontent.com/6101677/135541272-4b62029f-f7e6-4f6d-8d93-e63b8631c24c.png)

Reply:
I move highlevel coordination to `__init__` and now `make_widgets` and `make_gui`  are now analogous and orthogonal.